src/aggregation/bucket/range.rs

```diff
@@ -0,0 +1,437 @@
+    /// Note that this aggregation includes the from value and excludes the to value for each
+    /// range. Extra buckets will be created until the first to, and last from.
+    pub buckets: Vec<Range<f64>>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeBucketEntry {
+    range: Range<u64>,
+    bucket: SegmentBucketDataEntry,
+}
+
+/// The collector puts values from the fast field into the correct buckets and does a conversion to
+/// the correct datatype.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeCollector {
+    /// The buckets containing the aggregation data.
+    buckets: Vec<SegmentRangeBucketEntry>,
+    field_type: Type,
+}
+
+impl SegmentRangeCollector {
+    pub fn into_intermediate_agg_result(self) -> IntermediateBucketResult {
+        let field_type = self.field_type;
+        let buckets = self
+            .buckets
+            .into_iter()
+            .map(move |range_bucket| {
+                (
+                    range_to_key(&range_bucket.range, &field_type),
+                    range_bucket.bucket.into(),
+                )
+            })
+            .collect();
+
+        IntermediateBucketResult { buckets }
+    }
+
+    pub(crate) fn from_req(
+        req: &RangeAggregation,
+        sub_aggregation: &AggregationsWithAccessor,
+        field_type: Type,
+    ) -> crate::Result<Self> {
+        // The range input on the request is f64.
+        // We need to convert to u64 ranges, because we read the values as u64.
+        // The mapping from the conversion is monotonic so ordering is preserved.
+        let buckets = extend_validate_ranges(&req.buckets, &field_type)?
+            .iter()
+            .map(|range| {
+                Ok(SegmentRangeBucketEntry {
+                    range: range.clone(),
+                    bucket: SegmentBucketDataEntry::KeyCount(SegmentBucketDataEntryKeyCount {
+                        key: range_to_key(range, &field_type),
+                        doc_count: 0,
+                        values: None,
+                        sub_aggregation: SegmentAggregationResultsCollector::from_req(
+                            sub_aggregation,
+                        )?,
+                    }),
+                })
+            })
+            .collect::<crate::Result<_>>()?;
+
+        Ok(SegmentRangeCollector {
+            buckets,
+            field_type,
+        })
+    }
+
+    #[inline]
+    pub(crate) fn collect(
+        &mut self,
+        doc: DocId,
+        bucket_with_accessor: &BucketAggregationWithAccessor,
+    ) {
+        let val = bucket_with_accessor.accessor.get(doc);
+        self.collect_val(val, doc, bucket_with_accessor);
+    }
+
+    #[inline]
+    fn collect_val(
+        &mut self,
+        val: u64,
+        doc: DocId,
+        bucket_with_accessor: &BucketAggregationWithAccessor,
+    ) {
+        let bucket_pos = self
+            .buckets
+            .binary_search_by(|probe| match probe.range.contains(&val) {
```

Comment:
We are reaching "Early optimization is the root of all evil" territory I guess...
But binary search in an array of float should be much faster than binary search in an array of objects.

It comes with an arguable hit on intent readability and an improvement on proofreadability.

Reply:
That's what I thought, but my benchmark says otherwise, where it is consistently slower. I think mainly because now we need to manage two arrays. linear search does look quite quite good though, probably since it is more sensitive to cache optimizations.

```
running 5 tests
test aggregation::bucket::range::bench::bench_small_range_binary_search_only_float                                       ... bench:     132,952 ns/iter (+/- 26,978)
test aggregation::bucket::range::bench::bench_small_range_contains_binary_search                                         ... bench:     120,086 ns/iter (+/- 6,664)
test aggregation::bucket::range::bench::bench_small_range_contains_linear_search                                         ... bench:     151,267 ns/iter (+/- 4,616)
test aggregation::bucket::range::bench::bench_small_range_contains_linear_search_only_float                              ... bench:      94,946 ns/iter (+/- 33,240)
test aggregation::bucket::range::bench::bench_small_range_contains_linear_search_overlapping_buckets                     ... bench:     315,108 ns/iter (+/- 42,815)
```